source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+        {
+          // TODO: Add particles
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::sort_particles_in_subdomains_and_cells()
+    {
+      // There are three reasons why a particle is not in its old cell:
+      // It moved to another cell, to another domain or it left the mesh.
+      // Sort the particles accordingly and deal with them
+      std::multimap<types::LevelInd, Particle<dim> >            moved_particles_cell;
+      std::multimap<types::subdomain_id, Particle<dim> > moved_particles_domain;
+      std::multimap<types::LevelInd, Particle<dim> >            lost_particles;
+
+      // Find the cells that the particles moved to.
+      // Note that the iterator in the following loop is increased in a
+      // very particular way, because it is changed, if elements
+      // get erased. A change can result in invalid memory access.
+      typename std::multimap<types::LevelInd, Particle<dim> >::iterator   it;
+      for (it=particles.begin(); it!=particles.end();)
+        {
+          // If we know the particles old cell, check if it is still inside
+          if (it->first != std::make_pair(-1,-1))
+            {
+              const typename parallel::distributed::Triangulation<dim>::active_cell_iterator
+              old_cell (&(this->get_triangulation()), it->first.first, it->first.second);
+
+              try
+                {
+                  const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(old_cell, it->second.get_location());
+                  if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                    {
+                      ++it;
+                      continue;
+                    }
+                }
+              catch (...)
+                {}
+            }
+
+          // The particle is not in its old cell. Look for the new cell.
+          typename parallel::distributed::Triangulation<dim>::active_cell_iterator cell;
+          try
+            {
+              cell = (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), it->second.get_location())).first;
+            }
+          catch (GridTools::ExcPointNotFound<dim> &)
+            {
+              // If we can find no cell for this particle it has left the domain due
+              // to an integration error or an open boundary.
+              lost_particles.insert(*it);
+              particles.erase(it++);
+              continue;
+            }
+
+
+          // Reinsert the particle into our domain if we found its cell
+          // Mark it for MPI transfer otherwise
+          if (cell->is_locally_owned())
+            {
+              const types::LevelInd found_cell = std::make_pair(cell->level(),cell->index());
+              moved_particles_cell.insert(std::make_pair(found_cell, it->second));
+            }
+          else
+            moved_particles_domain.insert(std::make_pair(cell->subdomain_id(),it->second));
```

Comment:
Currently nothing. I will need to implement something, but as long as the CFL number is <= 1.0, this case should never happen (in theory). I will leave a todo.